ocaml/xapi/xapi_cluster.ml

```diff
@@ -108,103 +112,82 @@ let pool_create ~__context ~network ~cluster_stack ~token_timeout ~token_timeout
+    | None -> all_hosts
+    | Some master_ch ->
+      List.filter ((<>) master_ch) all_hosts
   in
-  debug "Destroying cluster_hosts in pool";
-  (* First try to destroy each cluster_host - if we can do so safely then do *)
-  List.iter
-    (fun cluster_host ->
-      (* We need to run this code on the slave *)
-      (* We ignore failures here, we'll try a force_destroy after *)
-      log_and_ignore_exn (fun () ->
-        Helpers.call_api_functions ~__context (fun rpc session_id ->
-            Client.Client.Cluster_host.destroy ~rpc ~session_id ~self:cluster_host)
-      )
-    )
-    slave_cluster_hosts;
-  (* We expect destroy to have failed for some, we'll try to force destroy those *)
-  (* Note we include the master here, we should attempt to force destroy it *)
-  let all_remaining_cluster_hosts =
-    Db.Cluster.get_cluster_hosts ~__context ~self
-  in
-  (* Now try to force_destroy, keep track of any errors here *)
-  let exns = List.fold_left
-    (fun exns_so_far cluster_host ->
-      Helpers.call_api_functions ~__context (fun rpc session_id ->
-        try
-          Client.Client.Cluster_host.force_destroy ~rpc ~session_id ~self:cluster_host;
-          exns_so_far
-        with e ->
-          Backtrace.is_important e;
-          let uuid = Client.Client.Cluster_host.get_uuid ~rpc ~session_id ~self:cluster_host in
-          debug "Ignoring exception while trying to force destroy cluster host %s: %s" uuid (ExnHelper.string_of_exn e);
-          e :: exns_so_far
-      )
-    )
-    [] all_remaining_cluster_hosts
-    in
+  foreach_cluster_host ~__context ~self ~log:force
+    ~fn:Client.Client.Cluster_host.destroy
+    slave_cluster_hosts
 
-    begin match exns with
-      | [] -> D.debug "Successfully destroyed all cluster_hosts in pool, now destroying cluster %s" (Ref.string_of self)
-      | e :: _ -> raise Api_errors.(Server_error (cluster_force_destroy_failed, [Ref.string_of self]))
-    end;
+let pool_force_destroy ~__context ~self =
+  (* Set pool_autojoin:false and try to destroy slave cluster_hosts *)
+  pool_destroy_common ~__context ~self ~force:true;
 
+  (* Note we include the master here, we should attempt to force destroy it *)
+  (* Now try to force_destroy, keep track of any errors here *)
+  debug "Ignoring exceptions while trying to force destroy cluster hosts.";
+  foreach_cluster_host ~__context ~self ~log:true
+    ~fn:Client.Client.Cluster_host.force_destroy
+    (Db.Cluster.get_cluster_hosts ~__context ~self);
+
+  info "Forgetting any cluster_hosts that couldn't be destroyed.";
+  foreach_cluster_host ~__context ~self ~log:true
+    ~fn:Client.Client.Cluster_host.forget
+    (Db.Cluster_host.get_all ~__context);
+
+  let unforgotten_cluster_hosts = List.filter
+      (fun self -> not (Db.Cluster_host.get_joined ~__context ~self))
+      (Db.Cluster_host.get_all ~__context)
+  in
+  info "If forget failed on any remaining cluster_hosts, we now delete them competely";
```

Comment:
This sentence is hard to parse. Should it be "If-forget failed .."?

Reply:
I'll add a small commit that fixes these to this PR.